File: evals/lib/records.ts
/**
 * The one place that knows the log schema. Both the eval runners and any later trajectory
 * debugger read turns through here, so raising `v` to 2 changes one file rather than every
 * consumer that happened to reach into a record.
 *
 * The types are restated here rather than imported from `src/agent/log.ts` because reader and
 * writer version independently: the Worker emits one schema version, a trace file on disk holds
 * every version ever written, and this side has to keep understanding the old ones after the
 * Worker has stopped producing them.
 */

export const SUPPORTED_SCHEMA_VERSION = 1

export type LogSource = 'telegram' | 'dev' | 'reflection' | 'reindex' | 'research' | 'schedule'

export interface AgentRecord {
  v: number
  ts: number
  level: 'info' | 'error'
  source: LogSource
  turnId: string
  seq: number
  at: string
  stage?: string
  round?: number
  roundsUsed?: number
  stopReason?: string
  outcome?: string
  subrequests?: number
  /** Present only when the deployment opted in; see LOG_CONTENT. */
  content?: { sample?: string; args?: string; resultHead?: string }
  [key: string]: unknown
}

export interface Turn {
  turnId: string
  source: LogSource
  /** Wall-clock of the first record; only meaningful for ordering turns against each other. */
  startedAt: number
  records: AgentRecord[]
  /** From the `turn` record, absent when the invocation died before writing one. */
  outcome?: string
  stopReason?: string
  subrequests?: number
  /** From the loop's `done` record. */
  roundsUsed?: number
  /** Whether a `turn` record was ever written. A false here is a finding, not a parse failure. */
  complete: boolean
}

/**
 * One record, two shapes. The observability *query* API flattens structured logs into top-level
 * keys; the *export* nests the whole object under `source`. Reading the wrong one looks exactly
 * like "the agent logged nothing", which it did not.
 *
 * Note the name collision: our envelope's own `source` (which entry point) sits inside
 * Cloudflare's `source` (the console call). Unwrapping keeps ours.
 */
export function fromExportedEvent(event: unknown): AgentRecord | null {
  if (typeof event !== 'object' || event === null) {
    return null
  }
  const outer = event as Record<string, unknown>
  if (isAgentRecord(outer)) {
    return outer
  }
  const nested = outer.source
  return isAgentRecord(nested) ? nested : null
}

export function isAgentRecord(value: unknown): value is AgentRecord {
  if (typeof value !== 'object' || value === null) {
    return false
  }
  const r = value as Record<string, unknown>
  return (
    r.v === SUPPORTED_SCHEMA_VERSION &&
    typeof r.turnId === 'string' &&
    r.turnId.length > 0 &&
    typeof r.seq === 'number' &&
    typeof r.at === 'string'
  )
}

/**
 * Groups records into turns. Anything without an envelope is dropped rather than pooled under a
 * synthetic id: an unattributable record is missing data, and inventing an id for it would make
 * it read as a turn that never happened.
 */
export function groupTurns(records: readonly AgentRecord[]): Turn[] {
  const byTurn = new Map<string, AgentRecord[]>()
  for (const record of records) {
    if (!isAgentRecord(record)) {
      continue
    }
    const bucket = byTurn.get(record.turnId)
    if (bucket) {
      bucket.push(record)
    } else {
      byTurn.set(record.turnId, [record])
    }
  }

  const turns: Turn[] = []
  for (const [turnId, unordered] of byTurn) {
    // seq is the order the code emitted them in; ts cannot be, since records inside one
    // invocation routinely share a millisecond.
    const ordered = [...unordered].sort((a, b) => a.seq - b.seq)
    const turnRecord = ordered.find((r) => r.at === 'turn')
    const done = ordered.find((r) => r.at === 'tool-loop' && r.stage === 'done')
    turns.push({
      turnId,
      source: ordered[0].source,
      startedAt: Math.min(...ordered.map((r) => r.ts)),
      records: ordered,
      outcome: turnRecord?.outcome as string | undefined,
      stopReason: (turnRecord?.stopReason ?? done?.stopReason) as string | undefined,
      subrequests: turnRecord?.subrequests as number | undefined,
      roundsUsed: done?.roundsUsed as number | undefined,
      complete: turnRecord !== undefined,
    })
  }

  return turns.sort((a, b) => a.startedAt - b.startedAt)
}

File: evals/lib/records.py
from collections import defaultdict
from dataclasses import dataclass, field
from typing import Any, TypeGuard, cast

SUPPORTED_SCHEMA_VERSION = 1

AgentRecord = dict[str, Any]


def as_record(value: object) -> AgentRecord | None:
    return cast(AgentRecord, value) if isinstance(value, dict) else None


@dataclass(frozen=True, slots=True)
class Turn:
    turn_id: str
    source: str
    started_at: int
    records: list[AgentRecord] = field(default_factory=list[AgentRecord])
    outcome: str | None = None
    stop_reason: str | None = None
    subrequests: int | None = None
    rounds_used: int | None = None
    complete: bool = False


def is_agent_record(value: object) -> TypeGuard[AgentRecord]:
    r = as_record(value)
    if r is None:
        return False
    turn_id = r.get('turnId')
    return (
        r.get('v') == SUPPORTED_SCHEMA_VERSION
        and isinstance(turn_id, str)
        and len(turn_id) > 0
        and isinstance(r.get('seq'), int | float)
        and isinstance(r.get('at'), str)
    )


def from_exported_event(event: object) -> AgentRecord | None:
    outer = as_record(event)
    if outer is None:
        return None
    if is_agent_record(outer):
        return outer
    nested = outer.get('source')
    return nested if is_agent_record(nested) else None


def group_turns(records: list[AgentRecord]) -> list[Turn]:
    by_turn: defaultdict[str, list[AgentRecord]] = defaultdict(list)
    for record in records:
        if is_agent_record(record):
            by_turn[record['turnId']].append(record)

    turns: list[Turn] = []
    for turn_id, unordered in by_turn.items():
        ordered = sorted(unordered, key=lambda r: r['seq'])
        turn_record = next((r for r in ordered if r['at'] == 'turn'), None)
        done = next((r for r in ordered if r['at'] == 'tool-loop' and r.get('stage') == 'done'), None)
        turns.append(
            Turn(
                turn_id=turn_id,
                source=ordered[0]['source'],
                started_at=min(r['ts'] for r in ordered),
                records=ordered,
                outcome=turn_record.get('outcome') if turn_record else None,
                stop_reason=(turn_record or {}).get('stopReason') or (done or {}).get('stopReason'),
                subrequests=turn_record.get('subrequests') if turn_record else None,
                rounds_used=done.get('roundsUsed') if done else None,
                complete=turn_record is not None,
            )
        )
    return sorted(turns, key=lambda t: t.started_at)
